xnmt/transformer.py

```diff
@@ -0,0 +1,315 @@
+        output = ReverseTimeDistributed()(output, seq_len, batch_size)
+    return output
+
+
+class LayerNorm(object):
+  def __init__(self, dy_model, d_hid):
+    self.p_g = dy_model.add_parameters(dim=d_hid)
+    self.p_b = dy_model.add_parameters(dim=d_hid)
+
+  def __call__(self, input_expr):
+    # g = dy.parameter(self.p_g)
+    # b = dy.parameter(self.p_b)
+    #
+    # (_, seq_len), batch_size = input_expr.dim()
+    # input = TimeDistributed()(input_expr)
+    # output = dy.layer_norm(input, g, b)
+    # return ReverseTimeDistributed()(output, seq_len, batch_size)
+    return input_expr
+
+
+def split_rows(X, h):
+  (n_rows, _), batch = X.dim()
+  l = range(n_rows)
+  steps = n_rows // h
+  output = []
+  for i in range(0, n_rows, steps):
+    output.append(dy.pickrange(X, i, i + steps))
+  return output
+
+
+def split_batch(X, h):
+  (n_rows, _), batch = X.dim()
+  l = range(batch)
+  steps = batch // h
+  output = []
+  for i in range(0, batch, steps):
+    indexes = l[i:i + steps]
+    output.append(dy.pick_batch_elems(X, indexes))
+  return output
+
+
+class MultiHeadAttention(object):
+  """ Multi Head Attention Layer for Sentence Blocks
+  For batch computation efficiency, dot product to calculate query-key
+  scores is performed all heads together.
+  """
+  def __init__(self, dy_model, n_units, h=1, self_attention=True, attn_dropout=False):
+    self.W_Q = Linear_nobias(dy_model, n_units, n_units)
+    self.W_K = Linear_nobias(dy_model, n_units, n_units)
+    self.W_V = Linear_nobias(dy_model, n_units, n_units)
+    self.finishing_linear_layer = Linear_nobias(dy_model, n_units, n_units)
+    self.h = h
+    self.scale_score = 1. / (n_units // h) ** 0.5
+    self.is_self_attention = self_attention
+    self.attn_dropout = attn_dropout
+
+  def set_dropout(self, dropout):
+    self.dropout = dropout
+
+  def __call__(self, x, z=None, mask=None):
+    h = self.h
+    if self.is_self_attention:
+      Q = self.W_Q(x)
+      K = self.W_K(x)
+      V = self.W_V(x)
+    else:
+      Q = self.W_Q(x)
+      K = self.W_K(z)
+      V = self.W_V(z)
+
+    (n_units, n_querys), batch = Q.dim()
+    (_, n_keys), _ = K.dim()
+
+    # Calculate Attention Scores with Mask for Zero-padded Areas
+    # Perform Multi-head Attention using pseudo batching
+    # all together at once for efficiency
+
+    batch_Q = dy.concatenate_to_batch(split_rows(Q, h))
+    batch_K = dy.concatenate_to_batch(split_rows(K, h))
+    batch_V = dy.concatenate_to_batch(split_rows(V, h))
+
+    assert(batch_Q.dim() == (n_units // h, n_querys), batch * h)
+    assert(batch_K.dim() == (n_units // h, n_keys), batch * h)
+    assert(batch_V.dim() == (n_units // h, n_keys), batch * h)
+
+    mask = np.concatenate([mask] * h, axis=0)
+    mask = dy.inputTensor(np.moveaxis(mask, [1, 0, 2], [0, 2, 1]), batched=True)
+    batch_A = (dy.transpose(batch_Q) * batch_K) * self.scale_score
+    batch_A = dy.cmult(batch_A, mask) + (1 - mask)*MIN_VALUE
```

Comment:
This is inefficient. I'd add `MIN_VALUE` to mask before calling `inputTensor`.

Reply:
mask is used in two places, in lines 139 and 147. I am actually not sure how to make this efficient. 
I have tried a couple of ways to do masking, but did not get noticeable increase in speed.